SSA/Core/Framework.lean

```diff
@@ -220,282 +590,717 @@ args, since there now are equation lemmas for it.
-    (e : Expr Op Γ ty)
-    (Γ'v : Valuation Γ') :
-    (e.changeVars varsMap).denote Γ'v =
-    e.denote (fun t v => Γ'v (varsMap v)) := by
+    (e : Expr Op Γ eff ty) :
+    (e.changeVars varsMap).denote =
+    fun Γ'v => e.denote (Γ'v.comap varsMap) := by
+  funext Γ'v
   rcases e with ⟨_, rfl, _⟩
-  simp [Expr.denote, Expr.changeVars, HVector.map_map]
+  simp [Expr.denote, Expr.changeVars, HVector.map_map, Valuation.comap]
+
+@[simp] lemma Com.changeVars_ret (v : Var Γ t) :
+    (Com.ret (Op:=Op) (eff := eff) v).changeVars = fun (map : Γ.Hom Δ) => Com.ret (map v) := by
+  funext map
+  simp [changeVars]
+
+@[simp] lemma Com.changeVars_lete (e : Expr Op Γ eff t) (body : Com Op _ eff u) :
+    (Com.lete e body).changeVars
+    = fun (map : Γ.Hom Δ) => Com.lete (e.changeVars map) (body.changeVars map.snocMap) := by
+  simp [changeVars]
+
+-- TODO: this is implied by simpler simp-lemmas, do we need it?
+@[simp] lemma Com.outContext_changeVars_ret (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+  ((Com.ret (Op:=Op) (eff := eff) v).changeVars varsMap).outContext = Γ' := by simp
+
+@[simp] lemma Com.denote_changeVars
+    (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+    (c.changeVars varsMap).denote =
+    c.denote ∘ (fun V => V.comap varsMap) := by
+  funext Γ'v
+  simp only [Function.comp_apply]
+  induction c using Com.rec' generalizing Γ'v Γ' with
+  | ret x => simp [Com.denote, Com.changeVars, Valuation.comap, *]
+  | lete _ _ ih =>
+    unfold denote
+    simp only [changeVars_lete]
+    split <;> simp_all
+
+@[simp] lemma Com.denote_changeVars' (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+    (c.changeVars varsMap).denote = fun V => c.denote (V.comap varsMap) := by
+  simp; rfl
+
+@[simp] lemma Com.outContext_changeVars_hom {map : Γ.Hom Δ} (map_inv : Δ.Hom Γ) :
+    {c : Com Op Γ eff ty} → Ctxt.Hom (outContext (changeVars c map)) (outContext c)
+  | .ret _        => cast (by simp) map_inv
+  | .lete _ body  => cast (by simp) <|
+      Com.outContext_changeVars_hom (map := map.snocMap) map_inv.snocMap (c := body)
+
+-- @[simp] lemma Com.denoteLets_changeVars (map : Γ.Hom Δ) (c : Com Op Γ .pure ty) :
+--     (c.changeVars map).denoteLets = fun (V : Valuation Δ) =>
+--       (c.denoteLets (V.comap map)).comap (fun t' v' => v'.castCtxt (by simp)) := by
+--   sorry
+
+@[simp] lemma Com.denoteLets_returnVar_pure (c : Com Op Γ .pure ty) (Γv : Valuation Γ) :
+    c.denoteLets Γv c.returnVar = c.denote Γv := by
+  induction c using Com.recPure <;> simp_all [denoteLets, denote]
+
+-- @[simp] lemma Com.comap_denoteLets_pure [LawfulMonad m]
+--     (map : Γ.Hom Δ) (c : Com Op Γ .pure ty) (Δv : Valuation Δ) :
+--     Valuation.comap ((c.changeVars map).denoteLets Δv)
+--       (c.changeVars map).outContextHom = Δv := by
+--   funext t' v'
+--   simp only [Valuation.comap]
+--   induction c using Com.recPure generalizing Δ Δv
+--   case ret => rfl
+--   case lete ih =>
+--     stop
+--     rw [changeVars_lete, outContextHom_lete];
+--     simp
```

Comment:
What's the issue with this lemma? Why is it commented out?

Reply:
Proof strategy that was not pursused. Removed for cleanliness.